docs/guides/streaming.md

```diff
@@ -0,0 +1,377 @@
+
+```tsx lines=[1,3,4,9-11,13-15,24-33,38-40]
+import { Suspense } from "react";
+import type { LoaderArgs } from "@remix-run/node"; // or cloudflare/deno
+import { defer } from "@remix-run/node"; // or cloudflare/deno
+import { Await, useLoaderData } from "@remix-run/react";
+
+import { getPackageLocation } from "~/models/packages";
+
+export function loader({ params }: LoaderArgs) {
+  const packageLocationPromise = getPackageLocation(
+    params.packageId
+  );
+
+  return defer({
+    packageLocation: packageLocationPromise,
+  });
+}
+
+export default function PackageRoute() {
+  const data = useLoaderData<typeof loader>();
+
+  return (
+    <main>
+      <h1>Let's locate your package</h1>
+      <Suspense
+        fallback={<p>Loading package location...</p>}
+      >
+        <Await
+          resolve={data.packageLocation}
+          errorElement={
+            <p>Error loading package location!</p>
+          }
+        >
+          {(packageLocation) => (
+            <p>
+              Your package is at {packageLocation.latitude}{" "}
+              lat and {packageLocation.longitude} long.
+            </p>
+          )}
+        </Await>
+      </Suspense>
+    </main>
+  );
+}
+```
+
+<details>
+  <summary>Alternatively, you can use the `useAsyncValue` hook:</summary>
+
+If you're not jazzed about bringing back render props, you can use a hook, but you'll have to break things out into another component:
+
+```tsx lines=[18,26-31]
+import type { UseDataFunctionReturn } from "@remix-run/react";
+
+export default function PackageRoute() {
+  const data = useLoaderData<typeof loader>();
+
+  return (
+    <main>
+      <h1>Let's locate your package</h1>
+      <Suspense
+        fallback={<p>Loading package location...</p>}
+      >
+        <Await
+          resolve={data.packageLocation}
+          errorElement={
+            <p>Error loading package location!</p>
+          }
+        >
+          <PackageLocation />
+        </Await>
+      </Suspense>
+    </main>
+  );
+}
+
+function PackageLocation() {
+  const packageLocation =
+    useAsyncValue<
+      UseDataFunctionReturn<
+        typeof loader
+      >["packageLocation"]
+    >();
+
+  return (
+    <p>
+      Your package is at {packageLocation.latitude} lat and{" "}
+      {packageLocation.longitude} long.
+    </p>
+  );
+}
+```
+
+</details>
+
+## Evaluating the solution
+
+So rather than waiting for the whole `document -> JavaScript -> hydrate -> request` cycle, with streaming we start the request for the slow data as soon as the document request comes in. This can significantly speed up the user experience.
+
+Additionally, the API that Remix exposes for this is extremely ergonomic. You can literally switch between whether something is going to be deferred or not based on whether you include the `await` keyword:
```

Comment:
```suggestion
Remix treats any `Promise` values as deferred data. These will be streamed to the client as each `Promise` resolves. If you'd like to prevent streaming for critical data, just use `await` and it will be included in the initial presentation of the document to the user.
```